internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+func (r *CloudEndpointReconciler) findTrafficPolicy(ctx context.Context, tpName, tpNamespace string) (string, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("name", tpName, "namespace", tpNamespace)
+
+	// Create a TrafficPolicy object to store the fetched result
+	tp := &ngrokv1alpha1.NgrokTrafficPolicy{}
+	key := client.ObjectKey{Name: tpName, Namespace: tpNamespace}
+
+	// Attempt to get the TrafficPolicy from the API server
+	if err := r.Client.Get(ctx, key, tp); err != nil {
+		log.Error(err, "failed to find TrafficPolicy CRD")
+		return "", err
+	}
+
+	// Convert the JSON policy to a string
+	policyBytes, err := tp.Spec.Policy.MarshalJSON()
+	if err != nil {
+		log.Error(err, "failed to marshal TrafficPolicy JSON")
+		return "", err
+	}
+
+	return string(policyBytes), nil
+}
+
+// ensureDomainExists checks if the Domain CRD exists, and if not, creates it.
+func (r *CloudEndpointReconciler) ensureDomainExists(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	domain := r.extractDomain(clep)
+	if domainEndsInReservedTLD(domain) {
+		// Skip creating the Domain CRD for reserved TLDs
+		return nil
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("domain", domain)
+
+	// Check if the Domain CRD already exists
+	domainObj := &ingressv1alpha1.Domain{}
+	err := r.Get(ctx, client.ObjectKey{Name: domain, Namespace: clep.Namespace}, domainObj)
+	if err == nil {
+		// Domain already exists
+		return nil
+	}
+	if client.IgnoreNotFound(err) != nil {
+		// Some other error occurred
+		log.Error(err, "failed to check Domain CRD existence")
+		return err
+	}
+
+	// Create the Domain CRD
+	newDomain := &ingressv1alpha1.Domain{
+		ObjectMeta: ctrl.ObjectMeta{
+			Name:      domain,
+			Namespace: clep.Namespace,
+		},
+		Spec: ingressv1alpha1.DomainSpec{
+			Domain: domain,
+		},
+	}
+	if err := r.Create(ctx, newDomain); err != nil {
+		log.Error(err, "failed to create Domain CRD")
+		return err
+	}
+
+	log.Info("Domain CRD created successfully, requeueing after delay")
+	return ErrDomainCreating
+}
+
+// domainEndsInReservedTLD checks if the domain ends in a reserved TLD (e.g., ".internal") in
+// order to filter it out of lists of domains to create automatically.
+func domainEndsInReservedTLD(domain string) bool {
+	// Check if the domain ends in the "internal" tld
+	return strings.HasSuffix(domain, ".internal")
+}
+
+// extractDomain parses the URL using Go's net/url package and extracts the host part.
+func (r *CloudEndpointReconciler) extractDomain(clep *ngrokv1alpha1.CloudEndpoint) string {
+	parsedURL, err := url.Parse(clep.Spec.URL)
+	if err != nil {
+		// Log the error and return an empty string if parsing fails
+		r.Log.Error(err, "failed to parse URL", "url", clep.Spec.URL)
+		return ""
+	}
+
+	// Extract the hostname from the parsed URL
+	host := parsedURL.Host
+
+	// Handle cases where the host may include a port (e.g., "example.com:8080")
+	if strings.Contains(host, ":") {
+		host = strings.Split(host, ":")[0]
+	}
+
```

Comment:
FWIW I found that `url.Parse()` then has `.Port()` that we can use.

Reply:
Looks like if i just use `parsedURL.Hostname()` instead it drops the port for me 👍